File: src/main/java/com/anod/appwatcher/sync/SyncTaskService.java
package com.anod.appwatcher.sync;

import android.content.ContentProviderClient;
import android.content.Intent;

import com.anod.appwatcher.AppListContentProvider;
import com.google.android.gms.gcm.GcmNetworkManager;
import com.google.android.gms.gcm.GcmTaskService;
import com.google.android.gms.gcm.TaskParams;

import info.anodsplace.android.log.AppLog;

public class SyncTaskService extends GcmTaskService {

    @Override
    public void onInitializeTasks() {
        super.onInitializeTasks();
        SyncScheduler.schedule(getApplicationContext());
    }

    @Override
    public int onRunTask(TaskParams taskParams) {
        AppLog.d("Scheduled call executed. Task: " + taskParams.getTag());

        SyncAdapter syncAdapter = new SyncAdapter(getApplicationContext());
        ContentProviderClient contentProviderClient = getContentResolver().acquireContentProviderClient(AppListContentProvider.AUTHORITY);

        int updatesCount = syncAdapter.onPerformSync(taskParams.getExtras(), contentProviderClient);

        Intent finishIntent = new Intent(SyncAdapter.SYNC_STOP);
        finishIntent.putExtra(SyncAdapter.EXTRA_UPDATES_COUNT, updatesCount);
        sendBroadcast(finishIntent);

        return GcmNetworkManager.RESULT_SUCCESS;
    }
}

File: src/main/java/com/anod/appwatcher/sync/SyncTaskService.kt
package com.anod.appwatcher.sync

import android.content.Intent
import android.os.Bundle
import com.anod.appwatcher.App
import com.anod.appwatcher.content.DbContentProvider
import com.firebase.jobdispatcher.JobParameters
import com.firebase.jobdispatcher.JobService
import info.anodsplace.android.log.AppLog
import info.anodsplace.appwatcher.framework.ApplicationContext
import info.anodsplace.appwatcher.framework.BackgroundTask

class SyncTaskService : JobService() {
    var runner: BackgroundTask<Void?, Int>? = null

    override fun onStartJob(job: JobParameters?): Boolean {
        if (job == null) return false

        AppLog.d("Scheduled call executed. Task: " + job.tag)

        this.runner = BackgroundTask(object : BackgroundTask.Worker<Void?, Int>(null) {
            override fun run(param: Void?): Int {
                val syncAdapter = UpdateCheck(ApplicationContext(applicationContext))
                val contentProviderClient = contentResolver.acquireContentProviderClient(DbContentProvider.authority)

                val extras = job.extras ?: Bundle.EMPTY
                return syncAdapter.perform(extras, contentProviderClient, App.log(applicationContext))
            }

            override fun finished(result: Int) {
                val finishIntent = Intent(UpdateCheck.syncStop)
                finishIntent.putExtra(UpdateCheck.extrasUpdatesCount, result)
                sendBroadcast(finishIntent)

                jobFinished(job, false)
            }
        })
        this.runner?.execute()
        return true
    }

    override fun onStopJob(job: JobParameters?): Boolean {
        AppLog.e("Job stopped. Task: ${job?.tag ?: "unknown"}")
        this.runner?.cancel(true)
        return true
    }
}



